shared/src/commonMain/kotlin/io/customer/shared/database/TrackingTaskQueryHelper.kt

```diff
@@ -0,0 +1,252 @@
+package io.customer.shared.database
+
+import com.squareup.sqldelight.TransactionWithReturn
+import io.customer.shared.runSuspended
+import io.customer.shared.sdk.config.BackgroundQueueConfig
+import io.customer.shared.sdk.meta.Workspace
+import io.customer.shared.tracking.constant.Priority
+import io.customer.shared.tracking.constant.QueueTaskStatus
+import io.customer.shared.tracking.model.*
+import io.customer.shared.tracking.queue.TaskResultListener
+import io.customer.shared.tracking.queue.failure
+import io.customer.shared.tracking.queue.success
+import io.customer.shared.util.DateTimeUtil
+import io.customer.shared.util.JsonAdapter
+import io.customer.shared.util.Logger
+import io.customer.shared.util.DatabaseUtil
+import kotlinx.datetime.Instant
+
+/**
+ * The class works as a bridge for SQL queries. All queries to database should be made using this
+ * class to provide abstraction and keep communication with database easier.
+ */
+internal interface TrackingTaskQueryHelper {
+    fun insertTask(
+        task: Task,
+        listener: TaskResultListener<Boolean>? = null,
```

Comment:
Why have a listener instead of using coroutines? I expected to see these functions be `suspend` functions.

Reply:
Mainly because the way we are designing BQ, we don't expect SDK wrappers to switch to background dispatchers for calling suspend functions. KMM will make it easy through listeners. e.g. the current queue works something like
```
val queueStatus = backgroundQueue.queueIdentifyProfile()
if (queueStatus) {
    // success
} else {
    // error
}
```
After KMM changes, it will be updated to something like
```
backgroundQueue.queueIdentifyProfile() { result ->
    if (result.isSuccess) {
        // success
    } else {
        // error
    }
}
```

But relooking at it, I think we can mark all methods in `TrackingTaskQueryHelper` as `suspend`, and expect `BackrgoundQueue` in KMM to switch to background instead. I'll need to think on this, thanks for pointing this out. I'll confirm after trying out the impact with this change.

However, the code sample shared above will remain the same to keep it easier for SDKs.

Reply:
Took some refactoring, but it looks better at the end. All methods in query helper are now suspended and callers will be taking responsibility for switching to background.

Reply:
> But relooking at it, I think we can mark all methods in TrackingTaskQueryHelper as suspend, and expect BackrgoundQueue in KMM to switch to background instead

I agree with this approach. 

FYI, KMM converts Kotlin `suspend` functions into Swift `async` functions so at least the iOS wrappers can call `suspend` nicely. However, I don't think the iOS Swift code will be calling `suspend` functions as the iOS SDK swift code will essentially be a wrapper around KMM code, but I still wanted to share this nice tip. 

> After KMM changes, it will be updated to something like
```
backgroundQueue.queueIdentifyProfile() { result ->
    if (result.isSuccess) {
        // success
    } else {
        // error
    }
}
```

I am not sure this will work. Here is why. 

All of our `public` SDK functions today are synchronous. How do you handle this? 

```swift
class CustomerIO {

  fun trackEvent(...) {
     // how do we call this async code from a synchronous function? 
     backgroundQueue.queueTrackEvent(...) { result ->  
     }
  }
}
```

Reply:
Thanks for the details on `async` functions for swift, that will be helpful. For the BQ interface, I mainly don't want to go synchronous because
- They might be called on main thread and will keep waiting till the task is inserted to database
- We are using database lock whenever updating task status to make sure everything works as expected, this can result in increased wait time if the database is already locked for some other queries

And to let synchronous functions communicate with asynchronous functions in the **BackgroundQueue**, they are currently designed like this
```
fun queueIdentifyProfile(profileIdentifier: String, attributes: CustomAttributes, listener: TaskResultListener<Unit>? = null)
```

Just for clarity, the methods in BQ are not suspended so they won't be async functions, instead, they will include an optional callback so callers can listen on the result if needed. The implementation of above method will look something like
```
fun queueIdentifyProfile(profileIdentifier: String, attributes: CustomAttributes, listener: TaskResultListener<Unit>? = null) {
    coroutineScope.launchOnBackground {
        val result = trackingTaskQueryHelper.insertTask()
        launchOnMain { listener?.onComplete(result) }
    }
}
```
Which means we will be switching between `background` and `main` in KMM as per our needs and expectations in the SDK. With this, [ProfileRepository in **Android**](https://github.com/customerio/customerio-android/blob/develop/sdk/src/main/java/io/customer/sdk/repository/ProfileRepository.kt#L43-L74) should look like
```
backgroundQueue.queueIdentifyProfile() { result ->
    if (result.isSuccess) {
        // success
    } else {
        // error
    }
}
```
Because `TaskResultListener` is [functional interface](https://kotlinlang.org/docs/fun-interfaces.html) and works well with Kotlin. But I'm yet to test this for iOS, but I'll give it a run and expecting [CustomerIOImplementation in **iOS**](https://github.com/customerio/customerio-ios/blob/develop/Sources/Tracking/CustomerIOImplementation.swift#L124-L153) to look much similar when calling BQ.

However, if this doesn't work well, we can probably use suspend functions with `async` for iOS and may be some good extensions or wrapper class for Android if needed? Haven't thought much about it but if the listener doesn't work well, I'll surely look for other options because I too want to keep `BackgroundQueue` calls easy for our SDKs.

I hope this explains my idea behind the BQ. We'll be able to test this once `BackgroundQueue` PR is available next week, and hope it works well for iOS too 🤞🏻

Reply:
I understand what you're saying about implementation of the API of the BQ. 

However, I am not yet understanding how we can provide a fully synchronous API to customers using our SDK. 

If all of our public SDK functions that customers use are synchronous, then I would expect that we also communicate with the BQ in a synchronous way (for convenience mostly?). 

But, that conversation can be had in another PR where we see a public SDK function calling the BQ. Then this conversation of implementation of the BQ will be more relevant.

Reply:
> How we can provide a fully synchronous API to customers using our SDK.

Mainly because we don't return values in apis exposed to the customers, so whether we do the operations synchronously or asynchronously, it shouldn't really impact them directly.

Agreed, this will be more relevant in the BQ PR, as we'll be able to see the exact code and test the impact. So let's continue this discussion once we have that PR available.